include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -0,0 +1,338 @@
+    }
+
+    // sort based on crowding distance
+    Info << "NSGA2::Optimize() Sort(Crowding Distance)" << std::endl;
+    std::sort(population.begin(),
+              population.end(),
+              [this, ranks, crowdingDistance, population](MatType candidateP,
+                                                          MatType candidateQ)
+              {
+                size_t idxP, idxQ;
+                for(int i = 0; i < population.size(); i++)
+                {
+                  if (arma::approx_equal(population[i], candidateP, "absdiff", epsilon))
+                  {
+                    idxP = i;
+                  }
+                  if (arma::approx_equal(population[i], candidateQ, "absdiff", epsilon))
+                  {
+                    idxQ = i;
+                  }
+                }
+
+                return CrowdingOperator(idxP,
+                                        idxQ,
+                                        ranks,
+                                        crowdingDistance);
+              }
+    );
+
+    // yeild new population P_{t+1}
+    Info << "NSGA2::Optimize() Get P(t+1)" << std::endl;
+    population.resize(populationSize);
+  }
+
+  std::vector<MatType> bestFront;
+
+  for(size_t f: fronts[0])
+  {
+    bestFront.push_back(population[f]);
+  }
+
+  Callback::EndOptimization(*this, objectives, iterate, callbacks...);
+  return bestFront;
+}
+
+template<typename MultiobjectiveFunctionType,
+         typename MatType>
+inline void NSGA2::EvaluateObjectives(std::vector<MatType> population,
+                                      MultiobjectiveFunctionType objectives,
+                                      std::vector<arma::vec>& calculatedObjectives)
+{
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    calculatedObjectives[i] = objectives.Evaluate(population[i]);
+  }
+}
+
+template<typename MatType>
+inline void NSGA2::BinaryTournamentSelection(std::vector<MatType>& population)
+{
+  std::vector<MatType> children;
+
+  while (children.size() < population.size())
+  {
+    size_t indexA = arma::randi<size_t>(arma::distr_param(0, populationSize - 1));
+    size_t indexB = arma::randi<size_t>(arma::distr_param(0, populationSize - 1));
+
+    if (indexA == indexB)
+    {
+      if (indexB < populationSize - 1)
+        indexB++;
+      else
+        indexB--;
+    }
+
+    MatType parentA = population[indexA];
+    MatType parentB = population[indexB];
+
+    MatType childA = parentA, childB = parentB;
+
+    Info << "NSGA2::BinaryTournamentSelection() Crossover" << std::endl;
+    Crossover(childA, childB, parentA, parentB);
+
+    Info << "NSGA2::BinaryTournamentSelection() Mutate(A)" << std::endl;
+    Mutate(childA);
+    Info << "NSGA2::BinaryTournamentSelection() Mutate(B)" << std::endl;
+    Mutate(childB);
+
+    children.push_back(childA);
+    children.push_back(childB);
+  }
+
+  population.reserve(population.size() + children.size());
```

Comment:
Could we reserve it to `2*populationSize` from the beginning?

Reply:
If there are an odd number of parents (say, n = 2k+1), the evolution procedure generates a 2n + 1 children (`while (children.size() <= population.size() {...}`). I think we should reserve `2 * populationSize + 1` for both cases (odd and even sized population) to be safe?